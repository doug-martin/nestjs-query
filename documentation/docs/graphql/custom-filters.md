---
title: Custom Filters
---

In addition to the built in filters, you can also define custom filtering operations.

There are 2 types of custom filters:

- Global type-based filters, that automatically work on all fields of a given GraphQL type.
- Custom entity-specific filters, that are custom-tailored for DTOs and do not require a specific backing field (more on
  that below).

[//]: # (TODO Add link to page)
:::important

This page describes how to implement custom filters at the GraphQL Level. The persistence layer needs to support them as
well. For now, only TypeOrm is implemented. See here:

- [TypeOrm Custom Filters](/docs/persistence/typeorm/custom-filters)

:::

## Global type-based filters

Type based filters are applied globally to all DTOs, based only on the underlying GraphQL Type.

### Extending the existing filters

Let's assume our persistence layer exposes a `isMultipleOf` filter which allows us to filter numeric fields and choose
only multiples of a user-supplied value. In order to expose that filter on all numeric GraphQL fields, we can do the
following in any typescript file (ideally this should run before the app is initialized):

```ts
import { registerTypeComparison } from '@nestjs-query/query-graphql';
import { IsBoolean, IsInt } from 'class-validator';
import { Float, Int } from '@nestjs/graphql';

registerTypeComparison(Number, 'isMultipleOf', { FilterType: Number, GqlType: Int, decorators: [IsInt()] });
registerTypeComparison(Int, 'isMultipleOf', { FilterType: Number, GqlType: Int, decorators: [IsInt()] });
registerTypeComparison(Float, 'isMultipleOf', { FilterType: Number, GqlType: Int, decorators: [IsInt()] });

// Note, this also works
// registerTypeComparison([Number, Int, Float], 'isMultipleOf', { FilterType: Number, GqlType: Int, decorators: [IsInt()] });
```

Where:

- FilterType is the typescript type of the filter
- GqlType is the GraphQL type that will be used in the schema
- decorators represents a list of decorators that will be applied to the filter class at the specific field used for the
  operation, used e.g. for validation purposes

The above snippet patches the existing Number/Int/Float FieldComparisons so that they expose the new
field/operation `isMultipleOf`. Example:

```graphql
input NumberFieldComparison {
    is: Boolean
    isNot: Boolean
    eq: Float
    neq: Float
    gt: Float
    gte: Float
    lt: Float
    lte: Float
    in: [Float!]
    notIn: [Float!]
    between: NumberFieldComparisonBetween
    notBetween: NumberFieldComparisonBetween
    isMultipleOf: Int
}
```

### Defining a filter on a custom scalar

Let's assume we have a custom scalar, to represent e.g. a geo point (i.e. {lat, lng}):

```ts
@Scalar('Point', (type) => Object)
export class PointScalar implements CustomScalar<any, any> {
  description = 'Point custom scalar type';

  parseValue(value: any): any {
    return { lat: value.lat, lng: value.lng };
  }

  serialize(value: any): any {
    return { lat: value.lat, lng: value.lng };
  }

  parseLiteral(ast: ValueNode): any {
    if (ast.kind === Kind.OBJECT) {
      return ast.fields;
    }
    return null;
  }
}
```

Now, we want to add a radius filter to all Point scalars. A radius filter is a filter that returns all entities whose
location is within a given distance from another point.

First, we need to define the filter type:

```ts
@InputType('RadiusFilter')
export class RadiusFilter {
  @Field(() => Number)
  lat!: number;

  @Field(() => Number)
  lng!: number;

  @Field(() => Number)
  radius!: number;
}
```

Then, we need to register said filter:

```ts
registerTypeComparison(PointScalar, 'distanceFrom', {
  FilterType: RadiusFilter,
  GqlType: RadiusFilter,
});
```

The above snippet creates a new comparison type for the Point scalar and adds the distanceFrom operations to it.
Example:

```graphql
input RadiusFilter {
    lat: Float!
    lng: Float!
    radius: Float!
}

input PointScalarFilterComparison {
    distanceFrom: RadiusFilter
}
```

Now, our persistence layer will be able to receive this new `distanceFrom` key for every property that is represented as
a Point scalar.

:::important

If the shape of the filter at the GraphQL layer is different from what the persistence layer expects, remember to use
an [Assembler and its convertQuery method!](/docs/concepts/advanced/assemblers#converting-the-query)

:::

### Disabling a type-based custom filter on specific fields of a DTO

Global filters are fully compatible with the [allowedComparisons](/docs/graphql/dtos/#example---allowedcomparisons)
option of the `@FilterableField` decorator.

## DTO-based custom filters

These custom filters are explicitly registered on a single DTO field, rather than at the type level. This can be useful
if the persistence layer exposes some specific filters only on some entities (e.g. "Filter all projects who more than 5
pending tasks" where we need to compute the number of pending tasks using a SQL sub-query in the where clause, instead
of having a computed field in the project entity).

:::important

DTO-based custom filters cannot be registered on existing DTO filterable fields, use type-based filters for that!

:::

In order to register a "pending tasks count" filter on our ProjectDto, we can do as follows:

```ts
registerDTOFieldComparison(TestDto, 'pendingTaskCount', 'gt', {
  FilterType: Number,
  GqlType: Int,
  decorators: [IsInt()],
});
```

Where:

- FilterType is the typescript type of the filter
- GqlType is the GraphQL type that will be used in the schema
- decorators represents a list of decorators that will be applied to the filter class at the specific field used for the
  operation, used e.g. for validation purposes

This will add a new operation to the GraphQL `TestDto` input type

```graphql
input TestPendingTaskCountFilterComparison {
  gt: Int
}

input TestDtoFilter {
  """
    ...Other fields defined in TestDTO
  """
  pendingTaskCount: TestPendingTaskCountFilterComparison
}
```

Now, graphQL will accept the new filter and our persistence layer will be able to receive the key `pendingTaskCount` for all filtering operations related to the "TestDto" DTO.
