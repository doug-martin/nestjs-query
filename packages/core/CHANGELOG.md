# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.23.0](https://github.com/doug-martin/nestjs-query/compare/v0.22.0...v0.23.0) (2021-02-26)

**Note:** Version bump only for package @nestjs-query/core





# [0.22.0](https://github.com/doug-martin/nestjs-query/compare/v0.21.2...v0.22.0) (2021-02-08)

**Note:** Version bump only for package @nestjs-query/core





## [0.21.2](https://github.com/doug-martin/nestjs-query/compare/v0.21.1...v0.21.2) (2020-10-23)


### Features

* **core:** added two new filter helpers ([031012e](https://github.com/doug-martin/nestjs-query/commit/031012e96bf99e1eb08c155059fd5106b38e9faf))





# [0.21.0](https://github.com/doug-martin/nestjs-query/compare/v0.20.2...v0.21.0) (2020-10-16)


### Bug Fixes

* **core:** Look up the proper assembler with inheritance ([8bd22c5](https://github.com/doug-martin/nestjs-query/commit/8bd22c5a40974c9011d0b472dc1ebe1328ba83f6))
* **eslint:** Fix eslint to recognize sub packages ([13fdd2b](https://github.com/doug-martin/nestjs-query/commit/13fdd2b31289dbc80316cbdb5aa32edbe596bad4))


### Features

* **mongoose:** Hardening reference support ([107bba0](https://github.com/doug-martin/nestjs-query/commit/107bba040a2b1d423deb4f1e428a43cecab48e79))
* **mongoose:** Switch to native mongoose support ([5cdfa39](https://github.com/doug-martin/nestjs-query/commit/5cdfa39b7d91cf0f8438ef3387a89aac850f4452))





## [0.20.2](https://github.com/doug-martin/nestjs-query/compare/v0.20.1...v0.20.2) (2020-10-01)


### Features

* **core:** parallelize queries within relation query service ([b339a2a](https://github.com/doug-martin/nestjs-query/commit/b339a2a9a3d1ad315d92eec67ab31af18617f6ca))





## [0.20.1](https://github.com/doug-martin/nestjs-query/compare/v0.20.0...v0.20.1) (2020-09-28)

**Note:** Version bump only for package @nestjs-query/core





# [0.20.0](https://github.com/doug-martin/nestjs-query/compare/v0.19.4...v0.20.0) (2020-09-17)


### Features

* **core:** Update assemblers to allow transforming create/update dtos ([5085d11](https://github.com/doug-martin/nestjs-query/commit/5085d1193a84396c9016821347c04f0e15eb04da))
* **core:** Update query service decorator to have correct generics ([74dc618](https://github.com/doug-martin/nestjs-query/commit/74dc618b61d1ce5575843accf5ea01066020f073))





## [0.19.4](https://github.com/doug-martin/nestjs-query/compare/v0.19.3...v0.19.4) (2020-09-15)


### Features

* **graphql:** Add keyset connections ([36bdbdd](https://github.com/doug-martin/nestjs-query/commit/36bdbdd9fda8b1db531ceb65c3a7c604c3da23fe))





## [0.19.3](https://github.com/doug-martin/nestjs-query/compare/v0.19.2...v0.19.3) (2020-09-09)


### Bug Fixes

* **graphql,hooks:** Allow getting hooks from parent classes ([59a0aeb](https://github.com/doug-martin/nestjs-query/commit/59a0aebc3dabd7d23ffde576a94bc588e768efbe))





## [0.19.2](https://github.com/doug-martin/nestjs-query/compare/v0.19.1...v0.19.2) (2020-09-03)

**Note:** Version bump only for package @nestjs-query/core





# [0.19.0](https://github.com/doug-martin/nestjs-query/compare/v0.18.1...v0.19.0) (2020-09-01)


### Features

* **core:** Update QueryService to allow additional filtering ([474369b](https://github.com/doug-martin/nestjs-query/commit/474369bd46ee82e3c8510f0564019627367d467c))





## [0.18.1](https://github.com/doug-martin/nestjs-query/compare/v0.18.0...v0.18.1) (2020-08-14)


### Bug Fixes

* **core:** Fix potential stack overflow with filter comparison ([f498802](https://github.com/doug-martin/nestjs-query/commit/f49880274a32e681d9072253135a8669bec7b3b2))


### Features

* **core:** refactor null compares and improve tests ([3582ed2](https://github.com/doug-martin/nestjs-query/commit/3582ed2f6b4aa5e3fa78bd9986621b9816566156))
* refactored filter builder to support nested object filters ([1ee8dbf](https://github.com/doug-martin/nestjs-query/commit/1ee8dbf5a0ae1a1258b203da1e68901e2b8d20f8))





# [0.18.0](https://github.com/doug-martin/nestjs-query/compare/v0.17.10...v0.18.0) (2020-08-11)

**Note:** Version bump only for package @nestjs-query/core





## [0.17.10](https://github.com/doug-martin/nestjs-query/compare/v0.17.9...v0.17.10) (2020-08-01)

**Note:** Version bump only for package @nestjs-query/core





## [0.17.8](https://github.com/doug-martin/nestjs-query/compare/v0.17.7...v0.17.8) (2020-07-28)


### Features

* **graphql:** Allow specifying allowed comparisons on filterable fields ([ced2792](https://github.com/doug-martin/nestjs-query/commit/ced27920e5c2278c2a04c027a692e25b3306f6cb))





## [0.17.7](https://github.com/doug-martin/nestjs-query/compare/v0.17.6...v0.17.7) (2020-07-27)


### Features

* **core:** Added applySort, applyPaging and applyQuery [#405](https://github.com/doug-martin/nestjs-query/issues/405) ([9f9ae0d](https://github.com/doug-martin/nestjs-query/commit/9f9ae0d0722c685483f1b2e1bd501a0f3df3ff85))





## [0.17.2](https://github.com/doug-martin/nestjs-query/compare/v0.17.1...v0.17.2) (2020-07-17)

**Note:** Version bump only for package @nestjs-query/core





# [0.17.0](https://github.com/doug-martin/nestjs-query/compare/v0.16.2...v0.17.0) (2020-07-16)


### Features

* **aggregations:** Add aggregations interfaces ([d67e733](https://github.com/doug-martin/nestjs-query/commit/d67e73393d2cb8d2f0dc131a8455bb798a270e14))
* **aggregations:** Add aggregations to graphql ([af075d2](https://github.com/doug-martin/nestjs-query/commit/af075d2e93b6abbbfbe32afcc917350f803fadaa))
* **aggregations,relations,core:** Add relation aggregation to core ([a489588](https://github.com/doug-martin/nestjs-query/commit/a4895881a1e9ff76811b264cc58eeea116b3edfd))





## [0.16.2](https://github.com/doug-martin/nestjs-query/compare/v0.16.1...v0.16.2) (2020-07-09)

**Note:** Version bump only for package @nestjs-query/core





# [0.16.0](https://github.com/doug-martin/nestjs-query/compare/v0.15.1...v0.16.0) (2020-07-05)


### Features

* **core:** Add type support for nest objects in filter ([cd9d0b5](https://github.com/doug-martin/nestjs-query/commit/cd9d0b524c1f4c384dc9e5ac6baeb5a49bc068e7))





# [0.15.0](https://github.com/doug-martin/nestjs-query/compare/v0.14.3...v0.15.0) (2020-06-23)


### Features

* **graphql,connection:** Add totalCount to connections ([ed1e84a](https://github.com/doug-martin/nestjs-query/commit/ed1e84a2feb6f89c3b270fcbc1d0eaf6aec5e575))





# [0.14.0](https://github.com/doug-martin/nestjs-query/compare/v0.13.2...v0.14.0) (2020-06-18)

**Note:** Version bump only for package @nestjs-query/core





# [0.13.0](https://github.com/doug-martin/nestjs-query/compare/v0.12.0...v0.13.0) (2020-06-12)

**Note:** Version bump only for package @nestjs-query/core





# [0.12.0](https://github.com/doug-martin/nestjs-query/compare/v0.11.8...v0.12.0) (2020-06-07)


### Features

* **graphql:** Add graphql subscriptions ([5dc987f](https://github.com/doug-martin/nestjs-query/commit/5dc987f435e0680192313e208359839f9c21d70b))





## [0.11.8](https://github.com/doug-martin/nestjs-query/compare/v0.11.7...v0.11.8) (2020-05-30)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.7](https://github.com/doug-martin/nestjs-query/compare/v0.11.6...v0.11.7) (2020-05-29)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.6](https://github.com/doug-martin/nestjs-query/compare/v0.11.5...v0.11.6) (2020-05-26)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.5](https://github.com/doug-martin/nestjs-query/compare/v0.11.4...v0.11.5) (2020-05-21)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.4](https://github.com/doug-martin/nestjs-query/compare/v0.11.3...v0.11.4) (2020-05-19)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.3](https://github.com/doug-martin/nestjs-query/compare/v0.11.2...v0.11.3) (2020-05-16)

**Note:** Version bump only for package @nestjs-query/core





## [0.11.2](https://github.com/doug-martin/nestjs-query/compare/v0.11.1...v0.11.2) (2020-05-14)


### Bug Fixes

* Fix lint issues ([c3407c0](https://github.com/doug-martin/nestjs-query/commit/c3407c0abfebe2ed6563cf754bab646af124a661))


### Features

* **graphql,core:** Add support for custom services and assemblers ([85e8658](https://github.com/doug-martin/nestjs-query/commit/85e8658c6acd495233cabb576c3458afcb8fff12))





## [0.11.1](https://github.com/doug-martin/nestjs-query/compare/v0.11.0...v0.11.1) (2020-05-11)


### Features

* **graphql:** Add support for auto-generated federations ([238f641](https://github.com/doug-martin/nestjs-query/commit/238f641967ea6668dfb7bd9034fec732da7fe38b))





# [0.11.0](https://github.com/doug-martin/nestjs-query/compare/v0.10.2...v0.11.0) (2020-05-09)


### Features

* **graphql:** Add graphql module ([282c421](https://github.com/doug-martin/nestjs-query/commit/282c421d0e6f67fe750fa6005f6cb7d960c8fbd0))





# [0.10.0](https://github.com/doug-martin/nestjs-query/compare/v0.9.0...v0.10.0) (2020-04-29)


### Features

* **sequelize:** Initial Sequelize support ([bfcf436](https://github.com/doug-martin/nestjs-query/commit/bfcf4368b96617113c0334cd78a8881e4952eb99))





# [0.9.0](https://github.com/doug-martin/nestjs-query/compare/v0.8.9...v0.9.0) (2020-04-26)

**Note:** Version bump only for package @nestjs-query/core





## [0.8.9](https://github.com/doug-martin/nestjs-query/compare/v0.8.8...v0.8.9) (2020-04-24)

**Note:** Version bump only for package @nestjs-query/core





## [0.8.7](https://github.com/doug-martin/nestjs-query/compare/v0.8.6...v0.8.7) (2020-04-23)

**Note:** Version bump only for package @nestjs-query/core
