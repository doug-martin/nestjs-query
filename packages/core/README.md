<p align="center">
  <a href="https://doug-martin.github.io/nestjs-query" target="blank"><img src="https://doug-martin.github.io/nestjs-query/img/logo.svg" width="120" alt="Nestjs-query Logo" /></a>
</p>

[![npm version](https://img.shields.io/npm/v/@nestjs-query/core.svg)](https://www.npmjs.org/package/@nestjs-query/core)
[![Test](https://github.com/doug-martin/nestjs-query/workflows/Test/badge.svg?branch=master)](https://github.com/doug-martin/nestjs-query/actions?query=workflow%3ATest+and+branch%3Amaster+)
[![Coverage Status](https://coveralls.io/repos/github/doug-martin/nestjs-query/badge.svg?branch=master)](https://coveralls.io/github/doug-martin/nestjs-query?branch=master)
[![Known Vulnerabilities](https://snyk.io/test/github/doug-martin/nestjs-query/badge.svg?targetFile=packages/core/package.json)](https://snyk.io/test/github/doug-martin/nestjs-query?targetFile=packages/core/package.json)

# `@nestjs-query/core`

Core package that contains interfaces for types used by [`@nestjs-query/query-graphql`](../query-graphql) and [`@nestjs-query/query-typeorm`](../query-typeorm)

## Installation

[Install Guide](https://doug-martin.github.io/nestjs-query/docs/introduction/install)

## Docs

* [Nestjs-query Docs](https://doug-martin.github.io/nestjs-query/docs/introduction/getting-started)


