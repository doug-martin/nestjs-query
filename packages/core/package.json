{
  "name": "@nestjs-query/core",
  "version": "0.17.8",
  "description": "Base query package",
  "author": "doug-martin <doug@dougamartin.com>",
  "homepage": "https://github.com/doug-martin/nestjs-query#readme",
  "license": "MIT",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "directories": {
    "lib": "dist/src",
    "test": "__tests__"
  },
  "files": [
    "dist/src/**"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/doug-martin/nestjs-query.git"
  },
  "scripts": {
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "build": "tsc -p tsconfig.build.json",
    "clean": "rm -rf ./dist && rm -rf tsconfig.tsbuildinfo"
  },
  "bugs": {
    "url": "https://github.com/doug-martin/nestjs-query/issues"
  },
  "dependencies": {
    "lodash.merge": "4.6.2"
  },
  "peerDependencies": {
    "@nestjs/common": "^7.0.0",
    "class-transformer": "^0.2.3 || ^0.3.0"
  },
  "devDependencies": {
    "@nestjs/common": "7.4.2",
    "@nestjs/testing": "7.4.2",
    "@types/lodash.merge": "4.6.6",
    "class-transformer": "0.3.1",
    "ts-mockito": "2.6.1",
    "typescript": "3.9.7"
  }
}
