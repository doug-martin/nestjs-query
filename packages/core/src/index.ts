/* eslint-disable import/export */
export * from './interfaces';
export * from './common';
export { InjectAssemblerQueryService, InjectQueryService, getQueryServiceToken } from './decorators';
export {
  QueryService,
  AssemblerQueryService,
  RelationQueryService,
  NoOpQueryService,
  QueryServiceRelation,
  ProxyQueryService,
} from './services';
export {
  transformFilter,
  transformQuery,
  transformSort,
  applyFilter,
  getFilterFields,
  QueryFieldMap,
  transformAggregateQuery,
  transformAggregateResponse,
  applySort,
  applyPaging,
  applyQuery,
  mergeQuery,
  mergeFilter,
  invertSort,
  getFilterComparisons,
  getFilterOmitting,
} from './helpers';
export {
  ClassTransformerAssembler,
  DefaultAssembler,
  AbstractAssembler,
  Assembler,
  AssemblerSerializer,
  AssemblerDeserializer,
  AssemblerFactory,
} from './assemblers';
export { NestjsQueryCoreModule, NestjsQueryCoreModuleOpts } from './module';
