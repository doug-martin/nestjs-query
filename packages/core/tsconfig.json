{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "./dist",
        "rootDir": "."
    },
    "include": [
        "./src",
        "./__tests__"
    ],
    "exclude": [
        "node_modules",
        "dist"
    ]
}
