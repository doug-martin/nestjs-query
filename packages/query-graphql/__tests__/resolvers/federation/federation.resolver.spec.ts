import { Class } from '@nestjs-query/core';
import { ObjectType, Query, Resolver, ID } from '@nestjs/graphql';
import { deepEqual, objectContaining, when } from 'ts-mockito';
import {
  CursorConnection,
  CursorQueryArgsType,
  FederationResolver,
  FilterableField,
  OffsetConnection,
  OffsetQueryArgsType,
  Relation,
  UnPagedRelation,
} from '../../../src';
import {
  generateSchema,
  createResolverFromNest,
  TestResolverDTO,
  TestService,
  TestRelationDTO,
} from '../../__fixtures__';

describe('FederationResolver', () => {
  const generateSDL = <DTO extends TestResolverDTO>(DTOClass: Class<DTO>): Promise<string> => {
    @Resolver(() => DTOClass)
    class TestSDLResolver extends FederationResolver(DTOClass) {
      @Query(() => DTOClass)
      test(): DTO {
        return { id: '1', stringField: 'foo' } as DTO;
      }
    }

    return generateSchema([TestSDLResolver]);
  };

  @ObjectType('TestFederated')
  @Relation('relation', () => TestRelationDTO)
  @Relation('custom', () => TestRelationDTO, { relationName: 'other' })
  @UnPagedRelation('unPagedRelations', () => TestRelationDTO)
  @OffsetConnection('relationOffsetConnection', () => TestRelationDTO)
  @CursorConnection('relationCursorConnection', () => TestRelationDTO)
  class TestFederatedDTO extends TestResolverDTO {
    @FilterableField(() => ID)
    id!: string;

    @FilterableField()
    stringField!: string;
  }

  @Resolver(() => TestFederatedDTO)
  class TestResolver extends FederationResolver(TestFederatedDTO) {
    constructor(service: TestService) {
      super(service);
    }
  }

  it('should not add federation methods if one and many are empty', async () => {
    const schema = await generateSDL(TestResolverDTO);
    expect(schema).toMatchSnapshot();
  });

  it('use the defined relations', async () => {
    const schema = await generateSDL(TestFederatedDTO);
    expect(schema).toMatchSnapshot();
  });

  describe('one', () => {
    describe('one relation', () => {
      it('should call the service findRelation with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver, TestFederatedDTO);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const output: TestRelationDTO = {
          id: 'id-2',
          testResolverId: dto.id,
        };
        when(
          mockService.findRelation(TestRelationDTO, 'relation', deepEqual([dto]), deepEqual({ filter: undefined })),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.findRelation(dto, {});
        return expect(result).toEqual(output);
      });

      it('should call the service findRelation with the provided dto and correct relation name', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver, TestFederatedDTO);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const output: TestRelationDTO = {
          id: 'id-2',
          testResolverId: dto.id,
        };
        when(
          mockService.findRelation(TestRelationDTO, 'other', deepEqual([dto]), deepEqual({ filter: undefined })),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.findCustom(dto, {});
        return expect(result).toEqual(output);
      });
    });
  });

  describe('many - connection', () => {
    describe('with cursor paging strategy', () => {
      it('should call the service findRelation with the provided dto', async () => {
        const { resolver, mockService } = await createResolverFromNest(TestResolver, TestFederatedDTO);
        const dto: TestResolverDTO = {
          id: 'id-1',
          stringField: 'foo',
        };
        const query: CursorQueryArgsType<TestRelationDTO> = {
          filter: { id: { eq: 'id-2' } },
          paging: { first: 1 },
        };
        const output: TestRelationDTO[] = [
          {
            id: 'id-2',
            testResolverId: dto.id,
          },
        ];
        when(
          mockService.queryRelations(
            TestRelationDTO,
            'relationCursorConnections',
            deepEqual([dto]),
            objectContaining({ ...query, paging: { limit: 2, offset: 0 } }),
          ),
        ).thenResolve(new Map([[dto, output]]));
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
        const result = await resolver.queryRelationCursorConnections(dto, query, {});
        return expect(result).toEqual({
          edges: [
            {
              cursor: 'YXJyYXljb25uZWN0aW9uOjA=',
              node: {
                id: output[0].id,
                testResolverId: dto.id,
              },
            },
          ],
          pageInfo: {
            endCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
            hasNextPage: false,
            hasPreviousPage: false,
            startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
          },
          totalCountFn: expect.any(Function),
        });
      });
    });
  });

  describe('with offset paging strategy', () => {
    it('should call the service findRelation with the provided dto', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver, TestFederatedDTO);
      const dto: TestResolverDTO = {
        id: 'id-1',
        stringField: 'foo',
      };
      const query: OffsetQueryArgsType<TestRelationDTO> = {
        filter: { id: { eq: 'id-2' } },
        paging: { limit: 1 },
      };
      const output: TestRelationDTO[] = [
        {
          id: 'id-2',
          testResolverId: dto.id,
        },
      ];
      when(
        mockService.queryRelations(
          TestRelationDTO,
          'relationOffsetConnections',
          deepEqual([dto]),
          objectContaining({ ...query, paging: { limit: 2 } }),
        ),
      ).thenResolve(new Map([[dto, output]]));
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.queryRelationOffsetConnections(dto, query, {});
      return expect(result).toEqual({
        nodes: output,
        pageInfo: { hasNextPage: false, hasPreviousPage: false },
        totalCountFn: expect.any(Function),
      });
    });
  });

  describe('with no paging strategy', () => {
    it('should call the service findRelation with the provided dto', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver, TestFederatedDTO);
      const dto: TestResolverDTO = {
        id: 'id-1',
        stringField: 'foo',
      };
      const query: OffsetQueryArgsType<TestRelationDTO> = {
        filter: { id: { eq: 'id-2' } },
        paging: { limit: 1 },
      };
      const output: TestRelationDTO[] = [
        {
          id: 'id-2',
          testResolverId: dto.id,
        },
      ];
      when(
        mockService.queryRelations(
          TestRelationDTO,
          'unPagedRelations',
          deepEqual([dto]),
          objectContaining({ filter: query.filter }),
        ),
      ).thenResolve(new Map([[dto, output]]));
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.queryUnPagedRelations(dto, query, {});
      return expect(result).toEqual(output);
    });
  });
});
