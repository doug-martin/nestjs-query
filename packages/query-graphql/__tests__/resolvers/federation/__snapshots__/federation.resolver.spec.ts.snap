// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FederationResolver should not add federation methods if one and many are empty 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`FederationResolver use the defined relations 1`] = `
type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type TestFederated {
  id: ID!
  stringField: String!
  relation: TestRelationDTO!
  custom: TestRelationDTO!
  unPagedRelations(
    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): [TestRelationDTO!]!
  relationOffsetConnections(
    """Limit or page results."""
    paging: OffsetPaging = {limit: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestFederatedRelationOffsetConnectionsConnection!
  relationCursorConnections(
    """Limit or page results."""
    paging: CursorPaging = {first: 10}

    """Specify to filter the records returned."""
    filter: TestRelationDTOFilter = {}

    """Specify to sort results."""
    sorting: [TestRelationDTOSort!] = []
  ): TestFederatedRelationCursorConnectionsConnection!
}

input TestRelationDTOFilter {
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

input TestRelationDTOSort {
  field: TestRelationDTOSortFields!
  direction: SortDirection!
  nulls: SortNulls
}

enum TestRelationDTOSortFields {
  id
  testResolverId
}

"""Sort Directions"""
enum SortDirection {
  ASC
  DESC
}

"""Sort Nulls Options"""
enum SortNulls {
  NULLS_FIRST
  NULLS_LAST
}

input OffsetPaging {
  """Limit the number of records returned"""
  limit: Int

  """Offset to start returning records from"""
  offset: Int
}

input CursorPaging {
  """Paginate before opaque cursor"""
  before: ConnectionCursor

  """Paginate after opaque cursor"""
  after: ConnectionCursor

  """Paginate first"""
  first: Int

  """Paginate last"""
  last: Int
}

"""Cursor for paging through collections"""
scalar ConnectionCursor

type TestRelationDTOEdge {
  """The node containing the TestRelationDTO"""
  node: TestRelationDTO!

  """Cursor for this node."""
  cursor: ConnectionCursor!
}

type PageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean

  """The cursor of the first returned record."""
  startCursor: ConnectionCursor

  """The cursor of the last returned record."""
  endCursor: ConnectionCursor
}

type TestFederatedRelationCursorConnectionsConnection {
  """Paging information"""
  pageInfo: PageInfo!

  """Array of edges."""
  edges: [TestRelationDTOEdge!]!
}

type OffsetPageInfo {
  """true if paging forward and there are more records."""
  hasNextPage: Boolean

  """true if paging backwards and there are more records."""
  hasPreviousPage: Boolean
}

type TestFederatedRelationOffsetConnectionsConnection {
  """Paging information"""
  pageInfo: OffsetPageInfo!

  """Array of nodes."""
  nodes: [TestRelationDTO!]!
}

type Query {
  test: TestFederated!
}

`;
